Market/Services/PurchaseService.cs

```diff
@@ -0,0 +1,222 @@
+﻿using AutoMapper;
+using Market.Data.Repositories.Interfaces;
+using Market.DTOs.Bill;
+using Market.DTOs.Product;
+using Market.DTOs.Purchase;
+using Market.DTOs.Users;
+using Market.Middleware;
+using Market.Migrations;
+using Market.Models;
+using Market.Services.Interfaces;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Options;
+
+namespace Market.Services
+{
+    public class PurchaseService : IPurchaseService
+    {
+        private readonly IProductRepository _productRepository;
+        private readonly IPurchaseRepository _purchaseRepository;
+        private readonly ILogger<PurchaseService> _logger;
+        private readonly IConfiguration _configuration;
+        private readonly IMapper _mapper;
+
+        public PurchaseService(IPurchaseRepository purchaseRepository, IProductRepository productRepository, IConfiguration configuration, ILogger<PurchaseService> logger, IMapper mapper)
+        {
+            _configuration = configuration ?? throw new ArgumentNullException(nameof(configuration));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _productRepository = productRepository ?? throw new ArgumentNullException(nameof(productRepository));
+            _purchaseRepository = purchaseRepository ?? throw new ArgumentNullException(nameof(_purchaseRepository));
+            _mapper = mapper ?? throw new ArgumentNullException(nameof(mapper));
+        }
+
+        public async Task<bool> NewPurchase(List<int> productsId, Purchase purchase)
+        {
+            decimal total = 0;
+            decimal subTotal = 0;
+
+            var products = IsExistProducts(productsId);
+
+            foreach (var product in products)
+            {
+                var purchaseProduct = new PurchaseProducts
+                {
+                    ProductId = product.Id,
+                    Purchase = purchase
+                };
+                subTotal += product.Price;
+                purchase.PurchaseProducts.Add(purchaseProduct);
+            }
+
+            if(purchase.DeliveryType == Purchase.Delivery.Express)
+            {
+                if(string.IsNullOrWhiteSpace(purchase.Address.AdditionalData)) throw new Exception("Additional Data of address is required");
```

Comment:
Use a specific exception instead of a general one. In this case, "ArgumentException".